### ISSUE TITLE
'secret_invalid' on verification

### ISSUE THREAD

**Issue by mtths** [FriendlyCaptcha/friendly-challenge#51]
we started to use captcha to protect certain searches on our website from abuse and since monday we see spikes of errors in our logs in response to our verification requests

response status: 401

    {"success":false,"errors":["secret_invalid"]}

this is right after a flurry of successful verifications: 
![Screenshot_2021-07-06 SearchTables - Grafana](https://user-images.githubusercontent.com/804633/124560707-67285580-de3d-11eb-8ae3-b8ba7979925a.png)

after about 10minutes everything is fine again (while the secret did not change).
we do not use site verification because quite a couple of domains are involved but we might give that a go. could it be that an attacker sends his own solutions and relies on us accepting those as per your documentations best practices?

**Comment by gzuidhof** [comment]
Hi mtths,

Thank you for letting us know. Around half a year we saw a similar issue (although it wasn't for the api secrets but for the configuration of a given sitekey), which had to do with our caching.

Right now our workers will cache the responses from the database for a given api key validation for 1200 seconds so that it doesn't have to incur the rountrip to the database on every request. My intuition tells me that somehow a bad value got in this cache (even though only succesful responses should get cached).

Steps I have taken:
* I have decreased the maximum cache duration to 2 minutes until we are sure this issue is resolved.
* I purged the existing cache.
* I added some additional sanity checks, and validation for the value from the cache.
* I made the way the cache key is calculated simpler (it's based on the hash, which may have lead to a weird UTF-8 string, but if anything that should just lead to cache misses if it is problematic).
* Added some logging so I can monitor it. I was unable to reproduce the issue locally.

Maybe it does the trick, but at this point it's hard to be sure. Do let me know if you keep seeing the same issue. I'll keep this issue open for now.

---

>  [...] could it be that an attacker sends his own solutions and relies on us accepting those as per your documentations best practices?

We check the API key separately from whether the API key is valid for the given solution. So provided you are actually sending a valid api key always, you should never see this specific error code. It doesn't matter what solution is sent, the secret gets verified the same way.

In other words if an attacker were to send puzzles that do not belong to your account they would be marked as invalid solutions instead.

**Comment by pepl** [comment]
Related to that, we see regular spikes of solved Captchas which very likely are done via humans. E.g. over 1000 solved puzzles every day between 06:00 and 06:10 and a second bunch around 15:00 our time zone:

![screenshot-ganapati geizhals at-2021 07 07-09_41_00](https://user-images.githubusercontent.com/118787/124719848-dcf8f380-df07-11eb-865e-785d01355bdb.png)

Any recommendations, e.g. regarding config, how to prevent those from passing?

**Comment by gzuidhof** [comment]
@mtths could you let me know if the issue shows itself again? It may be related: On cloudflare if many requests came in from a single origin, subsequent requests from our worker that runs on the edge (e.g. to our database) would sometimes get ratelimited. Cloudflare has now removed this rate limiting for our services.

@pepl That does seem like automated traffic, if you are on a business plan (or higher) you can enable dynamic difficulty in the app settings - which makes it so that users that are likely bots & those that sent a lot of requests in a short timespan get increasingly difficult puzzles. In this way instead of their machine taking 2s to solve a captcha, it may be xx seconds, which makes spamming less feasible.

Another thing you can do, regardless of your plan, is to increase the base difficulty value (which is the minimum puzzle difficulty your users will get). Depending on your audience (i.e. are they still using internet explorer 11?), as well as what kind of operations you have the captchas for (if its a long form then 5 seconds of solving wont hurt) may mean that your users won't even notice. You also do this from the app management page, it will also show estimates for how long it would take to solve.

In the future it would be great if we could dynamically estimate a good base difficulty for the user (i.e. by looking at the distribution of the amount of time passed since the puzzle was generated and when it was submitted), but for now that adjustment is manual.

**Comment by mtths** [comment]
@gzuidhof only one further error during the last 24 hours

today at 11:59:59.424 CEST

    {"success":false,"errors":["error_talking_to_solution_store"]}

so it looks so far like your fiddling helped.

note that we are still not using sitekeys for our requests/verifications, this would involve setting up around a dozen domains which is why we stopped there.

**Comment by gzuidhof** [comment]
Thank you for confirming :). The above error does seem to occur a few times per day (on hundreds of thousands of requests).

We  currently have a different database in shadow mode for this purpose, once we see that it is reliable and stable for a good few weeks we will turn it on in production as the main source of truth. This issue may show up intermittently until then, but this case shouldn't matter much for security: the puzzle has at that point been verified in every other way :).

I'm closing this issue for now, feel free to re-open if things change.
